### PR TITLE
Avoid unlink call on empty string

### PR DESCRIPTION
**Pull request by pkesseli**
Replaced move constructor by "swap" on default-constructed, empty
string. This because the std::string move constructor does not guarantee
that the moved string is actually erased. Also added test so we don't
unnecessarily call "unlink" on an empty string, even though this has no
effect.

### CODE REVIEW

File: src/util/tempfile.cpp [hunk 1]
```diff
@@ -164,5 +164,6 @@ Function: temporary_filet::~temporary_filet
 
 temporary_filet::~temporary_filet()
 {
-  unlink(name.c_str());
+  if(!name.empty())
+    unlink(name.c_str());
 }
```
Comment (kroening) [inline src/util/tempfile.cpp]: The if should not be needed any more; make that an assertion?

**Comment by pkesseli** [comment]
In the current implementation, the name would be empty in a temporary_filet that was moved from. This to indicate that the destructor shouldn't delete anything. If we want to keep this invariant of the name not being empty, I can introduce an additional bool flag that indicates that the object was moved from.

**Comment by tautschnig** [comment]
I'm  wondering whether it is reasonable to move a temporary file, but I guess that's the case. As such the current approach should be fine.

**Comment by pkesseli** [comment]
We deleted the copy constructor (which definitely doesn't make sense, since that would cause double deletes) and explicitly implemented this move constructor. The use case for this is to be able to return a temporary_filet, e.g. from a utility function.

**Pull request merged**